File: Colage/Features/List/ListDiscoveryView.swift
import SwiftUI

/// List discovery mode — 2-column grid sorted by distance with distance slider
struct ListDiscoveryView: View {
    @ObservedObject var students: NearbyStudentsViewModel
    @EnvironmentObject var appState: AppState
    @EnvironmentObject var universityService: UniversityService
    @State private var selectedStudent: NearbyStudent?

    private let columns = [
        GridItem(.flexible(), spacing: 12),
        GridItem(.flexible(), spacing: 12)
    ]

    var body: some View {
        ZStack {
            ColageColors.background.ignoresSafeArea()

            VStack(spacing: 0) {
                // Distance slider
                VStack(spacing: 8) {
                    HStack {
                        Text("Within")
                            .font(ColageFonts.caption)
                            .foregroundStyle(ColageColors.textSecondary)
                        Spacer()
                        Text(distanceLabel)
                            .font(ColageFonts.monoSmall)
                            .foregroundStyle(universityService.currentTheme?.primary ?? ColageColors.primary)
                    }

                    Slider(
                        value: $students.maxDistance,
                        in: 50...5000,
                        step: 50
                    )
                    .tint(universityService.currentTheme?.primary ?? ColageColors.primary)
                }
                .padding(.horizontal, 20)
                .padding(.top, 80)
                .padding(.bottom, 12)

                // Floor filter
                HStack(spacing: 8) {
                    FloorFilterChip(
                        label: "All Floors",
                        isSelected: students.filterFloor == nil,
                        action: { students.filterFloor = nil }
                    )

                    FloorFilterChip(
                        label: "Floor \(appState.currentFloor)",
                        isSelected: students.filterFloor == appState.currentFloor,
                        action: { students.filterFloor = appState.currentFloor }
                    )

                    Spacer()

                    Text("\(students.filteredStudents.count) nearby")
                        .font(ColageFonts.caption)
                        .foregroundStyle(ColageColors.textTertiary)
                }
                .padding(.horizontal, 20)
                .padding(.bottom, 12)

                // Student grid
                if students.filteredStudents.isEmpty {
                    Spacer()
                    VStack(spacing: 12) {
                        Image(systemName: "person.slash")
                            .font(.system(size: 40))
                            .foregroundStyle(ColageColors.textTertiary)
                        Text("No students within range")
                            .font(ColageFonts.body)
                            .foregroundStyle(ColageColors.textSecondary)
                        Text("Try increasing the distance")
                            .font(ColageFonts.caption)
                            .foregroundStyle(ColageColors.textTertiary)
                    }
                    Spacer()
                } else {
                    ScrollView {
                        LazyVGrid(columns: columns, spacing: 12) {
                            ForEach(students.filteredStudents) { student in
                                StudentCard(
                                    student: student,
                                    themeColor: universityService.currentTheme?.primary ?? ColageColors.primary
                                )
                                .onTapGesture {
                                    selectedStudent = student
                                }
                            }
                        }
                        .padding(.horizontal, 16)
                        .padding(.bottom, 100)
                    }
                }
            }
        }
        .sheet(item: $selectedStudent) { student in
            MiniProfileSheet(student: student)
                .presentationDetents([.fraction(0.35), .large])
                .presentationDragIndicator(.visible)
                .presentationBackgroundInteraction(.enabled)
        }
    }

    private var distanceLabel: String {
        students.maxDistance.formattedDistance
    }
}

// MARK: - Student Card

struct StudentCard: View {
    let student: NearbyStudent
    let themeColor: Color

    var body: some View {
        VStack(spacing: 10) {
            // Avatar
            AvatarView(
                imageURL: student.profile.profilePhotoURL,
                size: 72,
                borderColor: themeColor,
                initials: student.profile.displayName.initials
            )

            // Name
            Text(student.profile.displayName)
                .font(ColageFonts.bodyBold)
                .foregroundStyle(ColageColors.textPrimary)
                .lineLimit(1)

            // Major
            if let major = student.profile.major {
                Text(major)
                    .font(ColageFonts.caption)
                    .foregroundStyle(ColageColors.textSecondary)
                    .lineLimit(1)
            }

            // Distance + Floor
            HStack(spacing: 6) {
                Image(systemName: "location.fill")
                    .font(.system(size: 10))
                    .foregroundStyle(themeColor)
                Text(student.distance.formattedDistance)
                    .font(ColageFonts.monoSmall)
                    .foregroundStyle(ColageColors.textSecondary)

                if student.location.floor > 1 {
                    Text("· F\(student.location.floor)")
                        .font(ColageFonts.monoSmall)
                        .foregroundStyle(ColageColors.textTertiary)
                }
            }
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, 16)
        .padding(.horizontal, 12)
        .background(ColageColors.surface)
        .clipShape(RoundedRectangle(cornerRadius: 16))
        .overlay(
            RoundedRectangle(cornerRadius: 16)
                .strokeBorder(ColageColors.border, lineWidth: 0.5)
        )
    }

    // Using Double.formattedDistance extension
}

// MARK: - Floor Filter Chip

struct FloorFilterChip: View {
    let label: String
    let isSelected: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(label)
                .font(ColageFonts.captionBold)
                .foregroundStyle(isSelected ? ColageColors.textPrimary : ColageColors.textTertiary)
                .padding(.horizontal, 12)
                .padding(.vertical, 6)
                .background(isSelected ? ColageColors.primary.opacity(0.2) : ColageColors.surface)
                .clipShape(Capsule())
                .overlay(
                    Capsule()
                        .strokeBorder(isSelected ? ColageColors.primary.opacity(0.4) : ColageColors.border, lineWidth: 1)
                )
        }
    }
}

#Preview {
    ListDiscoveryView(students: {
        let vm = NearbyStudentsViewModel()
        vm.loadMockData()
        return vm
    }())
    .environmentObject(AppState())
    .environmentObject(UniversityService())
}

File: android/app/src/main/java/com/colageclub/colage/features/list/ListDiscoveryView.kt
package com.colageclub.colage.features.list

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.colageclub.colage.core.design.*
import com.colageclub.colage.data.models.NearbyStudent
import com.colageclub.colage.features.discovery.MiniProfileSheet
import com.colageclub.colage.features.discovery.NearbyStudentsViewModel
import kotlin.math.abs

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ListDiscoveryView(
    viewModel: NearbyStudentsViewModel,
    currentFloor: Int,
    themeColor: Color = ColageColors.Primary
) {
    val maxDistance by viewModel.maxDistance.collectAsState()
    val filterFloor by viewModel.filterFloor.collectAsState()
    val students = viewModel.filteredStudents()
    var selectedStudent by remember { mutableStateOf<NearbyStudent?>(null) }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(ColageColors.Background)
    ) {
        Spacer(Modifier.height(80.dp))

        // Distance slider
        Column(modifier = Modifier.padding(horizontal = 20.dp)) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                Text("Within", style = ColageFonts.Caption.copy(color = ColageColors.TextSecondary))
                Text(
                    viewModel.listDistanceFeet.formattedDistance(),
                    style = ColageFonts.MonoSmall.copy(color = themeColor)
                )
            }
            Slider(
                value = maxDistance,
                onValueChange = { viewModel.setMaxDistance(it) },
                valueRange = 0f..1f,
                colors = SliderDefaults.colors(
                    thumbColor = themeColor,
                    activeTrackColor = themeColor
                )
            )
        }

        // Floor filter chips
        Row(
            modifier = Modifier.padding(horizontal = 20.dp, vertical = 4.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            FloorChip("All Floors", filterFloor == null) { viewModel.setFilterFloor(null) }
            FloorChip("Floor $currentFloor", filterFloor == currentFloor) { viewModel.setFilterFloor(currentFloor) }
            Spacer(Modifier.weight(1f))
            Text("${students.size} nearby", style = ColageFonts.Caption.copy(color = ColageColors.TextTertiary))
        }

        Spacer(Modifier.height(12.dp))

        if (students.isEmpty()) {
            Box(
                modifier = Modifier.fillMaxSize(),
                contentAlignment = Alignment.Center
            ) {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Icon(Icons.Default.PersonOff, null, tint = ColageColors.TextTertiary, modifier = Modifier.size(40.dp))
                    Spacer(Modifier.height(12.dp))
                    Text("No students within range", style = ColageFonts.Body.copy(color = ColageColors.TextSecondary))
                    Text("Try increasing the distance", style = ColageFonts.Caption.copy(color = ColageColors.TextTertiary))
                }
            }
        } else {
            LazyVerticalGrid(
                columns = GridCells.Fixed(2),
                contentPadding = PaddingValues(horizontal = 16.dp, vertical = 8.dp),
                horizontalArrangement = Arrangement.spacedBy(12.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                items(students, key = { it.id }) { student ->
                    StudentCard(
                        student = student,
                        themeColor = themeColor,
                        onClick = { selectedStudent = student }
                    )
                }
            }
        }
    }

    // Bottom sheet
    selectedStudent?.let { student ->
        val sheetState = rememberModalBottomSheetState()
        ModalBottomSheet(
            onDismissRequest = { selectedStudent = null },
            sheetState = sheetState,
            containerColor = ColageColors.Background,
            dragHandle = { BottomSheetDefaults.DragHandle(color = ColageColors.Border) }
        ) {
            MiniProfileSheet(student = student, themeColor = themeColor)
        }
    }
}

@Composable
fun StudentCard(
    student: NearbyStudent,
    themeColor: Color,
    onClick: () -> Unit
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(ColageColors.Surface, RoundedCornerShape(16.dp))
            .border(0.5.dp, ColageColors.Border, RoundedCornerShape(16.dp))
            .clickable(onClick = onClick)
            .padding(vertical = 16.dp, horizontal = 12.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        AvatarView(
            imageUrl = student.profile.profilePhotoURL,
            size = 72.dp,
            borderColor = themeColor,
            initials = student.profile.displayName.initials()
        )
        Spacer(Modifier.height(10.dp))
        Text(
            student.profile.displayName,
            style = ColageFonts.BodyBold.copy(color = ColageColors.TextPrimary),
            maxLines = 1,
            overflow = TextOverflow.Ellipsis
        )
        student.profile.major?.let {
            Text(it, style = ColageFonts.Caption.copy(color = ColageColors.TextSecondary), maxLines = 1, overflow = TextOverflow.Ellipsis)
        }
        Spacer(Modifier.height(4.dp))
        Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
            Icon(Icons.Default.LocationOn, null, tint = themeColor, modifier = Modifier.size(10.dp))
            Text(student.distance.formattedDistance(), style = ColageFonts.MonoSmall.copy(color = ColageColors.TextSecondary))
            if (student.location.floor != 1) {
                Text(
                    "· ${if (student.location.floor < 0) "B${abs(student.location.floor)}" else "F${student.location.floor}"}",
                    style = ColageFonts.MonoSmall.copy(color = ColageColors.TextTertiary)
                )
            }
        }
    }
}

@Composable
fun FloorChip(label: String, isSelected: Boolean, onClick: () -> Unit) {
    Text(
        text = label,
        style = ColageFonts.CaptionBold.copy(
            color = if (isSelected) ColageColors.TextPrimary else ColageColors.TextTertiary
        ),
        modifier = Modifier
            .background(
                if (isSelected) ColageColors.Primary.copy(alpha = 0.2f) else ColageColors.Surface,
                RoundedCornerShape(50)
            )
            .border(
                1.dp,
                if (isSelected) ColageColors.Primary.copy(alpha = 0.4f) else ColageColors.Border,
                RoundedCornerShape(50)
            )
            .clickable(onClick = onClick)
            .padding(horizontal = 12.dp, vertical = 6.dp)
    )
}
